File: app/src/main/java/com/houseofyeti/howler/audio/Calibration.kt
package com.houseofyeti.howler.audio

/**
 * Maps measured dBFS → estimated dB SPL. See docs/howler-calibration-design.md.
 *
 * Math: SPL = measured_dBFS + offset.
 *  - tier-1 (device-reported): offset = 94 − sensitivity  (94 dB SPL reference)
 *  - tier-2 (manual single-point): offset = reference_SPL − measured_dBFS
 *
 * The resolver is the transferable asset — it picks the best source per device.
 * STEP ZERO showed tier-1 getSensitivity() can be populated-but-garbage, so a
 * validated manual offset wins, and implausible device values are rejected.
 */
sealed interface Calibration {
    /** No mapping; report relative dBFS only. Always available, never lies. */
    data object Uncalibrated : Calibration

    /** Tier-2 manual single-point against a reference meter. */
    data class Manual(val offsetDb: Double, val referenceMeterClass: String) : Calibration

    /** Tier-1 device-reported transducer sensitivity (dBFS @ 94 dB SPL, 1 kHz). */
    data class DeviceReported(val sensitivityDbFs: Double) : Calibration
}

private const val REFERENCE_SPL = 94.0

/** dB SPL for a measured dBFS, or null when uncalibrated. */
fun Calibration.splFromDbfs(dbfs: Float): Float? = when (this) {
    is Calibration.Uncalibrated -> null
    is Calibration.Manual -> (dbfs + offsetDb).toFloat()
    is Calibration.DeviceReported -> (dbfs + (REFERENCE_SPL - sensitivityDbFs)).toFloat()
}

/** Tier-2 offset from one calibration point: reference_SPL − measured_dBFS. */
fun manualOffset(referenceSpl: Double, measuredDbfs: Double): Double = referenceSpl - measuredDbfs

/** A transducer sensitivity is only believable in this band; outside it is garbage. */
fun isPlausibleSensitivity(sensitivityDbFs: Double): Boolean = sensitivityDbFs in -60.0..-20.0

/**
 * Pick the best calibration source. A stored manual offset (validated) beats the
 * device-reported value; a device value is used only if plausible; else uncalibrated.
 */
fun resolveCalibration(
    storedManual: Calibration.Manual?,
    deviceSensitivityDbFs: Double?,
): Calibration = when {
    storedManual != null -> storedManual
    deviceSensitivityDbFs != null && isPlausibleSensitivity(deviceSensitivityDbFs) ->
        Calibration.DeviceReported(deviceSensitivityDbFs)
    else -> Calibration.Uncalibrated
}

File: ios/HowlerMeter/HowlerMeter/Calibration.swift
import Foundation

/// Maps measured dBFS → estimated dB SPL. iOS mirror of the Android
/// `Calibration.kt` (see docs/howler-calibration-design.md).
///
/// Math: `SPL = dBFS + offset`, tier-2 manual single-point:
///   offset = reference_SPL − measured_dBFS
///
/// The Android tier-1 "device-reported" source (offset = 94 − sensitivity) is
/// intentionally omitted: iOS exposes no transducer-sensitivity API, and Android
/// never auto-trusts it at runtime either (MainActivity resolves with a null
/// device value), so manual-or-uncalibrated matches real behavior on both.
enum Calibration: Equatable {
    /// No mapping; report relative dBFS only. Always available, never lies.
    case uncalibrated
    /// Tier-2 manual single-point against a reference meter.
    case manual(offsetDb: Double, referenceMeterClass: String)

    /// dB SPL for a measured dBFS, or nil when uncalibrated.
    func splFromDbfs(_ dbfs: Float) -> Float? {
        switch self {
        case .uncalibrated: return nil
        case .manual(let offsetDb, _): return dbfs + Float(offsetDb)
        }
    }
}

/// Tier-2 offset from one calibration point: reference_SPL − measured_dBFS.
func manualOffset(referenceSpl: Double, measuredDbfs: Double) -> Double {
    referenceSpl - measuredDbfs
}

/// Pick the best calibration source: a stored manual point, else uncalibrated.
/// (Kept as a seam — if a future iOS sensitivity source appears, it slots here.)
func resolveCalibration(storedManual: Calibration?) -> Calibration {
    storedManual ?? .uncalibrated
}

/// Persists the tier-2 manual calibration for this install (one device profile,
/// v1) in UserDefaults. Mirror of Android `CalibrationStore` (SharedPreferences).
struct CalibrationStore {
    private let defaults: UserDefaults
    init(defaults: UserDefaults = .standard) { self.defaults = defaults }

    func loadManual() -> Calibration? {
        // Absence of the offset key = never calibrated (distinct from offset 0).
        guard defaults.object(forKey: Keys.offset) != nil else { return nil }
        let offset = defaults.double(forKey: Keys.offset)
        let meter = defaults.string(forKey: Keys.meter) ?? ""
        return .manual(offsetDb: offset, referenceMeterClass: meter)
    }

    func saveManual(offsetDb: Double, referenceMeterClass: String) {
        defaults.set(offsetDb, forKey: Keys.offset)
        defaults.set(referenceMeterClass, forKey: Keys.meter)
    }

    func clear() {
        defaults.removeObject(forKey: Keys.offset)
        defaults.removeObject(forKey: Keys.meter)
    }

    private enum Keys {
        static let offset = "howler_calibration.offset_db"
        static let meter = "howler_calibration.meter_class"
    }
}
